discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -38,13 +38,26 @@ impl ArticleFilter for DuplicateFilter {
         stack: &[Document],
         mut articles: Vec<Article>,
     ) -> Result<Vec<Article>, GenericError> {
-        let urls = history
+        let mut urls = history
             .iter()
-            .map(|doc| doc.url.as_str())
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
+            .map(|doc| Cow::from(doc.url.as_str()))
+            .chain(stack.iter().map(|doc| Cow::from(doc.resource.url.as_str())))
             .collect::<HashSet<_>>();
 
-        articles.retain(|article| !urls.contains(&article.link.as_str()));
+        let mut titles = history
+            .iter()
+            .map(|doc| Cow::from(&doc.title))
+            .chain(stack.iter().map(|doc| Cow::from(&doc.resource.title)))
+            .collect::<HashSet<_>>();
+
+        articles.retain(|article| {
+            let do_retain = !(urls.contains(&Cow::from(article.link.as_str()))
+                || titles.contains(&Cow::from(&article.title)));
```

Comment:
Not is missing for `titles.contains`. Can you please add a test for this case?

Reply:
the `!` is outside of the braces which enclose the `||`, so this should work as intended. we could rewrite the `!(url || titles)` to `!url && !titles` to make it easier to read.

Reply:
ah you are right. I thought the last `)` on the urls line was closing the not.